File: backend/app/services/conditional_logic.py
def is_empty_value(value):
    if value is None:
        return True
    if isinstance(value, str):
        return value.strip() == ""
    if isinstance(value, (list, tuple, set, dict)):
        return len(value) == 0
    return False


def normalize_text(value):
    if isinstance(value, bool):
        return "true" if value else "false"
    return str(value or "").strip().lower()


def as_number(value):
    try:
        if value is None or (isinstance(value, str) and value.strip() == ""):
            return None
        return float(value)
    except (TypeError, ValueError):
        return None


def comparison_list(value):
    if isinstance(value, (list, tuple, set)):
        return {normalize_text(item) for item in value if normalize_text(item) != ""}
    return {normalize_text(item) for item in str(value or "").split(",") if normalize_text(item) != ""}


def field_id(value):
    try:
        return int(value)
    except (TypeError, ValueError):
        return value


def compare(operator, answer, comparison):
    empty = is_empty_value(answer)
    if operator == "is_empty":
        return empty
    if operator == "is_not_empty":
        return not empty
    if empty:
        return False

    if operator in {"greater_than", "less_than"}:
        left = as_number(answer)
        right = as_number(comparison)
        if left is None or right is None:
            return False
        return left > right if operator == "greater_than" else left < right

    if operator in {"in", "not_in"}:
        allowed = comparison_list(comparison)
        if isinstance(answer, (list, tuple, set)):
            matched = any(normalize_text(item) in allowed for item in answer)
        else:
            matched = normalize_text(answer) in allowed
        return matched if operator == "in" else not matched

    if isinstance(answer, (list, tuple, set)):
        values = [normalize_text(item) for item in answer]
        target = normalize_text(comparison)
        if operator == "equals":
            return target in values
        if operator == "not_equals":
            return target not in values
        if operator == "contains":
            return any(target in item for item in values)
        return False

    left = normalize_text(answer)
    right = normalize_text(comparison)
    if operator == "equals":
        return left == right
    if operator == "not_equals":
        return left != right
    if operator == "contains":
        return right in left
    return False


def evaluate_conditional_rules(fields, rules, answers):
    fields = fields or []
    rules = rules or []
    answers = answers or {}
    states = {
        str(field.get("id")): {
            "visible": True,
            "required": bool(field.get("required")),
        }
        for field in fields
    }
    grouped = {}
    for rule in rules:
        target = str(rule.get("target_field_id"))
        if target not in states:
            continue
        grouped.setdefault(target, []).append(rule)

    for target, target_rules in grouped.items():
        show_rules = [rule for rule in target_rules if rule.get("action") == "show"]
        hide_rules = [rule for rule in target_rules if rule.get("action") == "hide"]
        require_rules = [rule for rule in target_rules if rule.get("action") == "require"]
        optional_rules = [rule for rule in target_rules if rule.get("action") == "optional"]

        def matches(rule):
            trigger = str(rule.get("trigger_field_id"))
            answer = answers.get(trigger, answers.get(field_id(trigger)))
            return compare(rule.get("operator"), answer, rule.get("comparison_value"))

        any_hide = any(matches(rule) for rule in hide_rules)
        any_show = any(matches(rule) for rule in show_rules)
        if any_hide:
            states[target]["visible"] = False
        elif show_rules:
            states[target]["visible"] = any_show
        else:
            states[target]["visible"] = True

        for rule in require_rules:
            if matches(rule):
                states[target]["required"] = True
        for rule in optional_rules:
            if matches(rule):
                states[target]["required"] = False
        if not states[target]["visible"]:
            states[target]["required"] = False

    return {"field_states": states}

File: frontend/src/utils/conditionalLogic.js
export const isEmptyValue=(value)=>{
  if(value===undefined||value===null)return true;
  if(typeof value==="string")return value.trim()==="";
  if(Array.isArray(value))return value.length===0;
  if(value instanceof File)return false;
  return false;
};

const text=value=>typeof value==="boolean"?(value?"true":"false"):String(value??"").trim().toLowerCase();
const number=value=>{
  if(value===undefined||value===null||String(value).trim()==="")return null;
  const parsed=Number(value);
  return Number.isFinite(parsed)?parsed:null;
};
const list=value=>new Set(String(value??"").split(",").map(text).filter(Boolean));

export function conditionMatches(rule,answers){
  const value=answers?.[rule.trigger_field_id]??answers?.[String(rule.trigger_field_id)];
  const operator=rule.operator;
  if(operator==="is_empty")return isEmptyValue(value);
  if(operator==="is_not_empty")return !isEmptyValue(value);
  if(isEmptyValue(value))return false;
  if(operator==="greater_than"||operator==="less_than"){
    const left=number(value);
    const right=number(rule.comparison_value);
    if(left===null||right===null)return false;
    return operator==="greater_than"?left>right:left<right;
  }
  if(operator==="in"||operator==="not_in"){
    const allowed=list(rule.comparison_value);
    const matched=Array.isArray(value)?value.some(item=>allowed.has(text(item))):allowed.has(text(value));
    return operator==="in"?matched:!matched;
  }
  if(Array.isArray(value)){
    const values=value.map(text);
    const target=text(rule.comparison_value);
    if(operator==="equals")return values.includes(target);
    if(operator==="not_equals")return !values.includes(target);
    if(operator==="contains")return values.some(item=>item.includes(target));
    return false;
  }
  const left=text(value);
  const right=text(rule.comparison_value);
  if(operator==="equals")return left===right;
  if(operator==="not_equals")return left!==right;
  if(operator==="contains")return left.includes(right);
  return false;
}

export function evaluateConditionalRules(fields=[],rules=[],answers={}){
  const states=Object.fromEntries(fields.map(field=>[String(field.id),{visible:true,required:!!field.required}]));
  const grouped=rules.reduce((map,rule)=>{
    const target=String(rule.target_field_id);
    if(!states[target])return map;
    map[target]=[...(map[target]||[]),rule];
    return map;
  },{});
  Object.entries(grouped).forEach(([target,targetRules])=>{
    const showRules=targetRules.filter(rule=>rule.action==="show");
    const hideRules=targetRules.filter(rule=>rule.action==="hide");
    const requireRules=targetRules.filter(rule=>rule.action==="require");
    const optionalRules=targetRules.filter(rule=>rule.action==="optional");
    const anyHide=hideRules.some(rule=>conditionMatches(rule,answers));
    const anyShow=showRules.some(rule=>conditionMatches(rule,answers));
    if(anyHide)states[target].visible=false;
    else if(showRules.length)states[target].visible=anyShow;
    else states[target].visible=true;
    requireRules.forEach(rule=>{if(conditionMatches(rule,answers))states[target].required=true});
    optionalRules.forEach(rule=>{if(conditionMatches(rule,answers))states[target].required=false});
    if(!states[target].visible)states[target].required=false;
  });
  return {field_states:states};
}
